src/storage/buffer_manager/buffer_manager.cpp

```diff
@@ -70,47 +65,85 @@ BufferManager::~BufferManager() = default;
+    while (true) {
+        auto currStateAndVersion = pageState->getStateAndVersion();
+        switch (PageState::getState(currStateAndVersion)) {
+        case PageState::EVICTED: {
+            if (pageState->tryLock(currStateAndVersion)) {
+                if (!claimAFrame(fileHandle, pageIdx, pageReadPolicy)) {
+                    pageState->unlock();
+                    throw BufferManagerException("Failed to claim a frame.");
+                }
+                return getFrame(fileHandle, pageIdx);
+            }
+        } break;
+        case PageState::UNLOCKED:
+        case PageState::MARKED: {
+            if (pageState->tryLock(currStateAndVersion)) {
+                return getFrame(fileHandle, pageIdx);
+            }
+        } break;
+        case PageState::LOCKED: {
+            continue;
+        }
+        }
+    }
+}
+
+void BufferManager::optimisticRead(BMFileHandle& fileHandle, common::page_idx_t pageIdx,
+    const std::function<void(uint8_t*)>& func) {
+    auto pageState = fileHandle.getPageState(pageIdx);
+    while (true) {
+        auto currStateAndVersion = pageState->getStateAndVersion();
+        switch (PageState::getState(currStateAndVersion)) {
+        case PageState::UNLOCKED: {
+            func(getFrame(fileHandle, pageIdx));
+            if (pageState->getStateAndVersion() == currStateAndVersion) {
+                return;
+            }
+        } break;
+        case PageState::MARKED: {
+            // If the page is marked, we try to switch to unlocked. If we succeed, we read the page.
+            if (pageState->tryClearMark(currStateAndVersion)) {
+                func(getFrame(fileHandle, pageIdx));
+                return;
+            }
+        } break;
+        case PageState::EVICTED: {
+            pin(fileHandle, pageIdx, PageReadPolicy::READ_PAGE);
+            unpin(fileHandle, pageIdx);
+        } break;
+        default: {
+            // When locked, continue the spinning.
+            continue;
+        }
         }
     }
-    auto retVal = getFrame(fileHandle, pageIdx);
-    fileHandle.releasePageLock(pageIdx);
-    return retVal;
 }
 
 // Important Note: The caller should make sure that they have pinned the page before calling this.
 void BufferManager::setPinnedPageDirty(BMFileHandle& fileHandle, page_idx_t pageIdx) {
-    fileHandle.acquirePageLock(pageIdx, LockMode::SPIN);
     auto pageState = fileHandle.getPageState(pageIdx);
-    if (pageState && pageState->getPinCount() >= 1) {
-        pageState->setDirty();
-        fileHandle.releasePageLock(pageIdx);
-    } else {
-        fileHandle.releasePageLock(pageIdx);
-        throw BufferManagerException("If a page is not in memory or is not pinned, cannot set "
-                                     "it to isDirty = true. filePath: " +
-                                     fileHandle.getFileInfo()->path +
-                                     " pageIdx: " + std::to_string(pageIdx) + ".");
+    while (true) {
```

Comment:
I think this while loop is not necessary. Similarly you should just set the state to dirty since a pinned page is guaranteed to be LOCKED. In fact, you might be able to just remove this function if there is a single line call you can make to change the state of a page to dirty. That is I see that you have an `->unlock()` call that does not take the existing state as an argument. You can have a `setLockedPageDirty()` function.

Reply:
Done. Also moved this function to bm_file_handle.